staging/src/k8s.io/apimachinery/pkg/util/proxy/websocket.go

```diff
@@ -0,0 +1,250 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package proxy
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/httpstream/spdy2"
+	"k8s.io/apimachinery/pkg/util/httpstream/websocket"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/klog/v2"
+)
+
+const (
+	stdinChannel = iota
+	stdoutChannel
+	stderrChannel
+	errorChannel
+	resizeChannel
+
+	preV4BinaryWebsocketProtocol = websocket.ChannelWebSocketProtocol
+	preV4Base64WebsocketProtocol = websocket.Base64ChannelWebSocketProtocol
+	v4BinaryWebsocketProtocol    = "v4." + websocket.ChannelWebSocketProtocol
+	v4Base64WebsocketProtocol    = "v4." + websocket.Base64ChannelWebSocketProtocol
+)
+
+// Options contains details about which streams are required for
+// remote command execution.
+type Options struct {
+	Stdin  bool
+	Stdout bool
+	Stderr bool
+	TTY    bool
+}
+
+// NewOptions creates a new Options from the Request.
+func streamOptionsFromRequest(req *http.Request) (*Options, error) {
+	query := req.URL.Query()
+	tty := query.Get("tty") == "true"
+	stdin := query.Get("stdin") == "true"
```

Comment:
I investigated the failing test `Pods should support remote command execution over websockets` in these jobs and sorted out the problem that;

`streamOptionsFromRequest` function expects query parameters in true/false format https://github.com/kubernetes/kubernetes/blob/19d128f55adfba9d8266c4b49c1e9ad71d300534/staging/src/k8s.io/apimachinery/pkg/util/proxy/websocket.go#L61C2-L62 but test sends in 1/0 https://github.com/kubernetes/kubernetes/blob/4d63e7d4d6d63109a3555c1011fe3b6c89bd2be6/test/e2e/common/node/pods.go#L566-L567. That's why, api server panicks.

When test sends true/false, it passes